File: backend/hackathon-backend/routers/analytics.py
"""
Analytics dashboard endpoints for reporting and insights
"""

from fastapi import APIRouter, Depends, HTTPException, Query
from typing import Dict, Any, Optional, List
from sqlalchemy.orm import Session

from database import get_db
from services.analytics_service import analytics_service
from dependencies import get_current_user
from models import User

router = APIRouter()


@router.get("/dashboard", response_model=Dict[str, Any])
async def get_dashboard_metrics(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Get comprehensive dashboard metrics
    Includes user activity, system usage, and project metrics
    """
    try:
        metrics = await analytics_service.get_metrics()
        return metrics
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error fetching metrics: {str(e)}")


@router.get("/user-activity", response_model=Dict[str, Any])
async def get_user_activity_metrics(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Get user activity metrics
    """
    try:
        metrics = await analytics_service.get_metrics("user_activity")
        return metrics
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error fetching user activity metrics: {str(e)}")


@router.get("/system-usage", response_model=Dict[str, Any])
async def get_system_usage_metrics(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Get system usage metrics from audit logs
    """
    try:
        metrics = await analytics_service.get_metrics("system_usage")
        return metrics
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error fetching system usage metrics: {str(e)}")


@router.get("/project-metrics", response_model=Dict[str, Any])
async def get_project_metrics(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Get project and deliverable metrics
    """
    try:
        metrics = await analytics_service.get_metrics("project_metrics")
        return metrics
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error fetching project metrics: {str(e)}")


@router.get("/user/{user_id}", response_model=Dict[str, Any])
async def get_user_analytics(
    user_id: int,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Get analytics for a specific user
    """
    # Only allow users to view their own analytics or admins
    if current_user.id != user_id and current_user.role != "admin":
        raise HTTPException(status_code=403, detail="Not authorized to view this user's analytics")
    
    try:
        user_analytics = await analytics_service.get_user_analytics(user_id)
        return user_analytics
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error fetching user analytics: {str(e)}")


@router.get("/health")
async def get_analytics_health():
    """
    Get analytics service health status
    """
    return {
        "status": "running" if analytics_service._is_running else "stopped",
        "last_updated": analytics_service.metrics_cache.get("timestamp") if analytics_service.metrics_cache else None
    }


@router.get("/health/public")
async def get_public_analytics_health():
    """
    Get public analytics service health status (no authentication required)
    """
    try:
        # Test that we can get metrics without authentication
        metrics = await analytics_service.get_metrics()
        return {
            "status": "running",
            "has_metrics": bool(metrics),
            "total_sprints": metrics.get("total_sprints", 0),
            "total_deliverables": metrics.get("total_deliverables", 0),
            "total_users": metrics.get("total_users", 0)
        }
    except Exception as e:
        return {
            "status": "error",
            "error": str(e)
        }


@router.get("/trends/daily-activity")
async def get_daily_activity_trend(
    days: int = Query(30, ge=1, le=365),
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Get daily activity trend for charting
    """
    if current_user.role != "admin":
        raise HTTPException(status_code=403, detail="Admin access required")
    
    try:
        from sqlalchemy import func, Date
        from datetime import datetime, timedelta
        
        daily_activity = db.query(
            func.date(analytics_service._AuditLog.timestamp).label("date"),
            func.count(analytics_service._AuditLog.id)
        ).filter(
            analytics_service._AuditLog.timestamp >= datetime.now() - timedelta(days=days)
        ).group_by(func.date(analytics_service._AuditLog.timestamp)).order_by("date").all()
        
        return [{"date": date.isoformat(), "count": count} for date, count in daily_activity]
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error fetching daily activity: {str(e)}")


@router.get("/trends/user-registration")
async def get_user_registration_trend(
    weeks: int = Query(12, ge=1, le=52),
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Get user registration trend for charting
    """
    if current_user.role != "admin":
        raise HTTPException(status_code=403, detail="Admin access required")
    
    try:
        from sqlalchemy import func
        from datetime import datetime, timedelta
        
        weekly_registrations = db.query(
            func.date_trunc('week', User.created_at).label("week"),
            func.count(User.id)
        ).filter(
            User.created_at >= datetime.now() - timedelta(weeks=weeks)
        ).group_by(func.date_trunc('week', User.created_at)).order_by("week").all()
        
        return [{"week": week.isoformat(), "count": count} for week, count in weekly_registrations]
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error fetching registration trend: {str(e)}")

File: backend/node-backend/src/routes/analytics.js
const express = require('express');
const router = express.Router();
const analyticsService = require('../services/analyticsService');
const { authenticateToken, requireRole } = require('../middleware/auth');

/**
 * @route GET /api/analytics/dashboard
 * @desc Get comprehensive dashboard metrics
 * @access Private
 */
router.get('/dashboard', authenticateToken, async (req, res) => {
  try {
    const metrics = await analyticsService.getMetrics();
    res.json(metrics);
  } catch (error) {
    console.error('Error fetching dashboard metrics:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

/**
 * @route GET /api/analytics/user-activity
 * @desc Get user activity metrics
 * @access Private
 */
router.get('/user-activity', authenticateToken, async (req, res) => {
  try {
    const metrics = await analyticsService.getMetrics('user_activity');
    res.json(metrics);
  } catch (error) {
    console.error('Error fetching user activity metrics:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

/**
 * @route GET /api/analytics/system-usage
 * @desc Get system usage metrics
 * @access Private
 */
router.get('/system-usage', authenticateToken, async (req, res) => {
  try {
    const metrics = await analyticsService.getMetrics('system_usage');
    res.json(metrics);
  } catch (error) {
    console.error('Error fetching system usage metrics:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// Compatibility endpoint for performance metrics expected by frontend
router.get('/performance', authenticateToken, async (req, res) => {
  try {
    const metrics = await analyticsService.getMetrics('system_usage');
    res.json(metrics);
  } catch (error) {
    console.error('Error fetching performance metrics:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

/**
 * @route GET /api/analytics/project-metrics
 * @desc Get project and deliverable metrics
 * @access Private
 */
router.get('/project-metrics', authenticateToken, async (req, res) => {
  try {
    const metrics = await analyticsService.getMetrics('project_metrics');
    res.json(metrics);
  } catch (error) {
    console.error('Error fetching project metrics:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

/**
 * @route GET /api/analytics/user/:user_id
 * @desc Get analytics for a specific user
 * @access Private
 */
router.get('/user/:user_id', authenticateToken, async (req, res) => {
  try {
    const { user_id } = req.params;
    
    // Only allow users to view their own analytics or admins
    if (req.user.id !== parseInt(user_id) && req.user.role !== 'admin') {
      return res.status(403).json({
        error: 'Not authorized',
        message: 'Not authorized to view this user\'s analytics'
      });
    }
    
    const userAnalytics = await analyticsService.getUserAnalytics(parseInt(user_id));
    res.json(userAnalytics);
  } catch (error) {
    console.error('Error fetching user analytics:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

/**
 * @route GET /api/analytics/health
 * @desc Get analytics service health status
 * @access Private
 */
router.get('/health', authenticateToken, async (req, res) => {
  try {
    res.json({
      status: analyticsService._isRunning ? 'running' : 'stopped',
      last_updated: analyticsService.metricsCache?.timestamp || null
    });
  } catch (error) {
    console.error('Error fetching analytics health:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

/**
 * @route GET /api/analytics/health/public
 * @desc Get public analytics service health status (no authentication required)
 * @access Public
 */
router.get('/health/public', async (req, res) => {
  try {
    // Test that we can get metrics without authentication
    const metrics = await analyticsService.getMetrics();
    res.json({
      status: 'running',
      has_metrics: !!metrics,
      total_sprints: metrics.total_sprints || 0,
      total_deliverables: metrics.total_deliverables || 0,
      total_users: metrics.total_users || 0
    });
  } catch (error) {
    res.json({
      status: 'error',
      error: error.message
    });
  }
});

/**
 * @route GET /api/analytics/trends/daily-activity
 * @desc Get daily activity trend for charting
 * @access Private (Admin only)
 */
router.get('/trends/daily-activity', authenticateToken, requireRole('admin'), async (req, res) => {
  try {
    const { days = 30 } = req.query;
    
    const dailyActivity = await AuditLog.findAll({
      attributes: [
        [Sequelize.fn('DATE', Sequelize.col('timestamp')), 'date'],
        [Sequelize.fn('COUNT', Sequelize.col('id')), 'count']
      ],
      where: {
        timestamp: {
          [Sequelize.Op.gte]: new Date(Date.now() - parseInt(days) * 24 * 60 * 60 * 1000)
        }
      },
      group: [Sequelize.fn('DATE', Sequelize.col('timestamp'))],
      order: [[Sequelize.fn('DATE', Sequelize.col('timestamp')), 'ASC']]
    });
    
    const result = dailyActivity.map(item => ({
      date: item.get('date').toISOString().split('T')[0],
      count: parseInt(item.get('count'))
    }));
    
    res.json(result);
  } catch (error) {
    console.error('Error fetching daily activity:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

/**
 * @route GET /api/analytics/trends/user-registration
 * @desc Get user registration trend for charting
 * @access Private (Admin only)
 */
router.get('/trends/user-registration', authenticateToken, requireRole('admin'), async (req, res) => {
  try {
    const { weeks = 12 } = req.query;
    
    const weeklyRegistrations = await User.findAll({
      attributes: [
        [Sequelize.fn('DATE_TRUNC', 'week', Sequelize.col('created_at')), 'week'],
        [Sequelize.fn('COUNT', Sequelize.col('id')), 'count']
      ],
      where: {
        created_at: {
          [Sequelize.Op.gte]: new Date(Date.now() - parseInt(weeks) * 7 * 24 * 60 * 60 * 1000)
        }
      },
      group: [Sequelize.fn('DATE_TRUNC', 'week', Sequelize.col('created_at'))],
      order: [[Sequelize.fn('DATE_TRUNC', 'week', Sequelize.col('created_at')), 'ASC']]
    });
    
    const result = weeklyRegistrations.map(item => ({
      week: item.get('week').toISOString().split('T')[0],
      count: parseInt(item.get('count'))
    }));
    
    res.json(result);
  } catch (error) {
    console.error('Error fetching registration trend:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

module.exports = router;